compiler/rustc_builtin_macros/src/deriving/debug.rs

```diff
@@ -197,15 +197,18 @@ fn show_substructure(cx: &mut ExtCtxt<'_>, span: Span, substr: &Substructure<'_>
 }
 
 /// Special case for enums with no fields. Builds:
+///
 /// ```text
-/// impl ::core::fmt::Debug for A {
+/// impl ::core::fmt::Debug for Fieldless {
 ///     fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
-///          ::core::fmt::Formatter::write_str(f,
-///             match self {
-///                 A::A => "A",
-///                 A::B() => "B",
-///                 A::C {} => "C",
-///             })
+///         ::core::fmt::Formatter::write_str(
+///             f,
+///             const { ["A", "B", "C"] }[match self {
+///                 Fieldless::A => 0usize,
+///                 Fieldless::B => 1usize,
+///                 Fieldless::C => 2usize,
+///             }],
```

Comment:
I have been looking into a similar case recently as noted in https://github.com/rust-lang/rust/issues/114106#issuecomment-1657131825, and part of the problem might be the `match` statement itself.

For data-less enums with *dense*/*contiguous* discriminants, you should be able to just use something like `*self as u32` to use the discriminant directly as an index instead of via the `match` roundtrip.